parlai/core/opt.py

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+Opt is the system for passing around options throughout ParlAI.
+"""
+
+import copy
+import json
+import pickle
+import traceback
+
+
+class Opt(dict):
+    """
+    Class for tracking options.
+
+    Functions like a dict, but allows us to track the history of arguments as they are
+    set.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.history = {}
+        self.deepcopies = []
+
+    def __setitem__(self, key, val):
+        loc = traceback.format_stack()[-2]
+        self.history.setdefault(key, []).append((loc, val))
+        super().__setitem__(key, val)
+
+    def __getstate__(self):
+        return (self.history, self.deepcopies, dict(self))
+
+    def __setstate__(self, state):
+        self.history, self.deepcopies, data = state
+        self.update(data)
+
+    def __reduce__(self):
+        return (Opt, (), self.__getstate__())
+
+    def __deepcopy__(self, memo):
+        """
+        Override deepcopy so that history is copied over to new object.
+        """
+        # track location of deepcopy
+        loc = traceback.format_stack()[-3]
```

Comment:
Hmm - why is it using the third-to-last call of the stack here?

Reply:
Three levels:
- `copy.deepcopy(opt)` (-3)
- inside the source of deepcopy it calls `opt__deepcopy__` (-2)
- here (-1)

We want to record the line that called deepcopy.